framework/calico_framework.py

```diff
@@ -0,0 +1,673 @@
+                    anywhere_task = anywhere_tasks.pop()
+                    anywhere_task.slave_id = offer.offer.slave_id.value
+
+                    tasks_for_offer.setdefault(offer, [])
+                    tasks_for_offer[offer].append(anywhere_task)
+                    reserved_offers_by_slave_id[offer.slave_id] = offer
+
+            # Check that all anywhere_tasks were assigned
+            if anywhere_tasks:
+                raise NotEnoughResources(
+                    "Matched all slave-specific tasks, but "
+                    "not enough remaining resources to launch "
+                    "the remaining %d anywhere-tasks" % len(anywhere_tasks))
+
+        except NotEnoughResources as e:
+            # Rollback all reservations
+            _log.debug("\t\tNot Launching: %s" % e)
+            for task in self.tasks:
+                task.slave_id = None
+            return {}
+
+        return reserved_offers_by_slave_id
+
+    def print_report(self):
+        """
+        Print this test's results.
+        """
+        print "|--- %s ---|" % self.name
+        failed_tasks = self.tasks_with_state(BAD_TASK_STATES)
+        if self.killed:
+            print "Test Status: KILLED"
+        elif failed_tasks:
+            print "Test Status: FAIL"
+        elif self.tasks_with_state([None]) == self.tasks:
+            print "Test Status: UNLAUNCHED"
+        else:
+            print "Test Status: PASS"
+
+        print "Reason: ", self.additional_info
+        print "Task Statuses:"
+        for task in self.tasks:
+            if task.state is None:
+                state = "UNLAUNCHED"
+            else:
+                state = mesos_pb2.TaskState.Name(task.state)
+            print "\t%s: %s" % (task, state)
+            if type(task) == PingTask:
+                print "\t\tTest Results: %s" % task.ping_status_data
+
+        print "-----------------------------------"
+
+    def tasks_with_state(self, states):
+        return [task for task in self.tasks if task.state in states]
+
+    def running(self):
+        if self.killed:
+            return False
+        else:
+            return self.tasks_with_state(
+                [mesos_pb2.TASK_RUNNING, mesos_pb2.TASK_STAGING]) != []
+
+    def launch(self, offers_by_slave_id):
+        """
+        Launches the test on the available offers.
+        This assumes the test has already reserved offers via self.can_run_on,
+        and assumes each task has already picked a task.slave_id.
+        """
+        operations_by_offer = {}
+        for task in self.tasks:
+            if not task.dependencies_are_met or task.slave_id not in offers_by_slave_id:
+                continue
+
+            offer = offers_by_slave_id[task.slave_id]
+
+            if offer not in operations_by_offer:
+                operation = mesos_pb2.Offer.Operation()
+                operation.type = mesos_pb2.Offer.Operation.LAUNCH
+                operations_by_offer[offer] = operation
```

Comment:
I think this block of code fails to define `operation` if `offer` is in `operations_by_offer`.  The usual Python pattern would be
```
try:
    operation = operations_by_offer[offer]
except KeyError:
    operation = mesos_pb2.Offer.Operation()
    operation.type = mesos_pb2.Offer.Operation.LAUNCH
    operations_by_offer[offer] = operation
```